### ISSUE TITLE
Add dns authenticator for FreeDNS

### ISSUE THREAD

**Issue by Duffyx** [certbot/certbot#6202]
It would be nice to have support for FreeDNS (https://freedns.afraid.org/) if possible.

**Comment by TomBZombie** [comment]
I quickly looked into this as it's something I'd like to do as well. Although FreeDNS has a  nice API for updating A records to your current IP, there isn't a simple API for updating a TXT record.  I emailed the owner a few days ago and don't yet have a reply.

It would be possible with curl (or similar) but all forms use internal IDs so you'd need to

- Log in using the log in form and store the cookie (easy)
- Parse the domain list page
      - Get the internal `domain_id`   for $domain
     -  Get `data_id`   for `_acme-challenge.$domain`, if it exists
- If it's a new record (no _acme-challenge TXT yet)  load `http://freedns.afraid.org/subdomain/edit.php?edit_domain_id=$domain_id` and extract the generated form token `ref`
- If it's an existing record, load `http://freedns.afraid.org/subdomain/edit.php?data_id=$data_id` and extract the generated form token `ref`
- post `data_id=$data_id' (if set) + `domain_id=$domain_Id&ref=$ref&send=Save!&subdomain=_acme-challenge&ttl=&type=TXT`  to `http://freedns.afraid.org/subdomain/save.php?step=2`


I could easily do this with JavaScript or PHP but my python knowledge is next to zero.

**Comment by ibaiul** [comment]
@TRPB Please update your response to include the https protocol in the example URL otherwise your credentials would be leaked in plain text over the network. `https://freedns.afraid.org/subdomain/save.php?step=2`

Here you can get a working example of the auth-hook script for dns authenticator:
```
#!/bin/bash

# Strip only the top domain to get the zone id
echo "Domain1: $CERTBOT_DOMAIN"
DOMAIN=$(expr match "$CERTBOT_DOMAIN" '.*\.\(.*\..*\)')
echo "Domain2: $DOMAIN"
TTL=60

echo "Validation: $CERTBOT_VALIDATION"
curl -b "dns_cookie=$YOUR_COOKIE" -d "type=TXT" -d "subdomain=_acme-challenge" -d "domain_id=$YOUR_DOMAIN_ID" -d "data_id=$YOUR_DATA_ID" -d "address=%22$CERTBOT_VALIDATION%22" -d "ttl=$TTL" https://freedns.afraid.org/subdomain/save.php?step=2

# Sleep to make sure the change has time to propagate over to DNS
sleep $TTL

echo "Output: $CERTBOT_AUTH_OUTPUT"
```

**Comment by seii** [comment]
Thanks to @ibaiul for providing a base, I was clueless about how the auth hooks were scripted. I'm going to take the liberty of expanding on the example, because it didn't seem to cover some of my own use cases. Unfortunately I also don't know python, so I wasn't able to explore making a true plugin.

My Environment:
- Existing Let's Encrypt certificate, but it was expired
- Multiple domains, multiple sub-domains
- Port 80 blocked by ISP, so DNS challenge was required

Notes:
- **DO NOT** use the `$DATA_ID` of one of your "A" records. You _will_ accidentally turn it into a "TXT" record and be confused like I was!
- **DO** use the `$DATA_ID` of some pre-created TXT records. These would have the "Subdomain" format of `_acme-challenge` (for domains without a subdomain, like `example.com`) or `_acme-challenge.$SUBDOMAIN` (for domains that do have a subdomain, like `www.example.com`)
- Using the below auth hook I was able to run this command successfully: `sudo certbot renew --manual --preferred-challenges=dns --manual-auth-hook=/some/directory/manual-auth-hook.sh --manual-public-ip-logging-ok`
- The above command is now run via a cronjob (minus "sudo") and seems to work just fine there too
- For some reason the use of "curl" makes certbot say there is "error output" after every command. You'll know if you were successful because you'll only see curl's progress indicator in the "error output". True errors are when you start seeing HTML, because that's FreeDNS displaying an error page
- I don't claim to be the most elegant Bash scripter ever, so apologies for any ugliness

manual-auth-hook.sh:

```
#!/bin/bash

DNS_COOKIE=<YOUR_DNS_COOKIE>

echo "Full Domain: $CERTBOT_DOMAIN"

# Get top domain (e.g. "example.com" from "www.example.com")
DOMAIN=$(expr match "$CERTBOT_DOMAIN" '.*\.\(.*\..*\)')

# 'expr match' returns nothing if there's no sub-domain
#    (e.g. "example.com"), so run an alternate search if that happens
if [ -z "$DOMAIN" ]
then
  DOMAIN=$(expr match "$CERTBOT_DOMAIN" '\(.*\..*\)')
fi
echo "Domain: $DOMAIN"

# Get sub-domain (e.g. "www" from "www.example.com")
SUBDOMAIN=$(expr match "$CERTBOT_DOMAIN" '\(.*\)\..*\..*')
if [ -z "$SUBDOMAIN" ]
then
  SUBDOMAIN="(none)"
fi
echo "Sub-domain: $SUBDOMAIN"

# Set time used to sleep and wait for DNS to catch up (in minutes)
TTL=60

# Variable examples for different types of domain
case "$CERTBOT_DOMAIN" in
	"example.com")
	   DOMAIN_ID=<YOUR_DOMAIN_ID>
	   DATA_ID=<YOUR_DATA_ID>
	   SUBDOMAIN_ID="_acme-challenge"
	   ;;
	"www.example.com")
	   DOMAIN_ID=<YOUR_DOMAIN_ID>
	   DATA_ID=<YOUR_DATA_ID>
	   SUBDOMAIN_ID="_acme-challenge.$SUBDOMAIN"
	   ;;
	"something.else.example.com")
	   DOMAIN_ID=<YOUR_DOMAIN_ID>
	   DATA_ID=<YOUR_DATA_ID>
	   SUBDOMAIN_ID="_acme-challenge.$SUBDOMAIN"
	   ;;
esac

echo "Attempting to use validation token $CERTBOT_VALIDATION with $CERTBOT_DOMAIN"
echo "DOMAIN_ID = $DOMAIN_ID"
echo "SUBDOMAIN ID = $SUBDOMAIN_ID"
echo "DATA_ID = $DATA_ID"
curl -b "dns_cookie=$DNS_COOKIE" -d "type=TXT" -d "subdomain=$SUBDOMAIN_ID" -d "domain_id=$DOMAIN_ID" -d "data_id=$DATA_ID" -d "address=%22$CERTBOT_VALIDATION%22" https://freedns.afraid.org/subdomain/save.php?step=2

# Sleep to make sure the change has time to propagate over to DNS
sleep $TTL

echo "Output: $CERTBOT_AUTH_OUTPUT"
```

**Comment by celsian** [comment]
@seii Your script is glorious. Thank you.

**Comment by TRPB** [comment]
for these scripts where do you get YOUR_DNS_COOKIE from?

**Comment by TRPB** [comment]
Thanks @seii, I've created a docker image here: https://hub.docker.com/repository/docker/trpb/certbot-freedns/general ( source: https://github.com/TRPB/docker-certbot-freedns ) and some instructions, took me a while to work out where to get the DNS_COOKIE value from.
